pandas/_libs/join.pyx

```diff
@@ -677,10 +677,10 @@ ctypedef fused by_t:
     uint64_t
 
 
-def asof_join_backward_on_X_by_Y(ndarray[asof_t] left_values,
-                                 ndarray[asof_t] right_values,
-                                 ndarray[by_t] left_by_values,
-                                 ndarray[by_t] right_by_values,
+def asof_join_backward_on_X_by_Y(asof_t[:] left_values,
+                                 asof_t[:] right_values,
```

Comment:
Can the asof_t ones be const?

Reply:
I tried but got compilation errors. I believe it's because of a cython limitation with const fused types as described in this open [issue](https://github.com/cython/cython/issues/1772) from the cython repo.